clippy_lints/src/manual_let_else.rs

```diff
@@ -0,0 +1,269 @@
+                    does_diverge = true;
+                    return false;
+                }
+                true
+            },
+            ExprKind::MethodCall(..) => {
+                if is_never(cx, ex) {
+                    does_diverge = true;
+                    return false;
+                }
+                true
+            },
+            ExprKind::If(if_expr, if_then, if_else) => {
+                let else_diverges = if_else.map_or(false, |ex| expr_diverges(cx, ex));
+                let diverges = expr_diverges(cx, if_expr) || (else_diverges && expr_diverges(cx, if_then));
+                if diverges {
+                    does_diverge = true;
+                }
+                false
+            },
+            ExprKind::Match(match_expr, match_arms, _) => {
+                let diverges =
+                    expr_diverges(cx, match_expr) || match_arms.iter().all(|arm| expr_diverges(cx, arm.body));
+                if diverges {
+                    does_diverge = true;
+                }
+                false
+            },
+
+            // Don't continue into loops, as they are breakable,
+            // and we'd have to start checking labels.
+            ExprKind::Loop(..) => false,
+
+            // Default: descend
+            // TODO: support breakable blocks. For this we would need label support.
+            _ => true,
+        }
+    })
+    .visit_expr(expr);
+    does_diverge
+}
+
+/// Returns true if the two spans come from different macro sites,
+/// or one comes from an invocation and the other is not from a macro at all.
+fn from_different_macros(span_a: Span, span_b: Span) -> bool {
+    // This pre-check is a speed up so that we don't build outer_expn_data unless needed.
+    match (span_a.from_expansion(), span_b.from_expansion()) {
+        (false, false) => return false,
+        (true, false) | (false, true) => return true,
+        // We need to determine if both are from the same macro
+        (true, true) => (),
+    }
+    let data_for_comparison = |sp: Span| {
+        let expn_data = sp.ctxt().outer_expn_data();
+        (expn_data.kind, expn_data.call_site)
+    };
+    data_for_comparison(span_a) != data_for_comparison(span_b)
+}
+
+fn pat_allowed_for_else(cx: &LateContext<'_>, pat: &'_ Pat<'_>) -> bool {
+    // Check whether the pattern contains any bindings, as the
+    // binding might potentially be used in the body.
+    // TODO: only look for *used* bindings.
+    let mut has_bindings = false;
+    pat.each_binding_or_first(&mut |_, _, _, _| has_bindings = true);
+    if has_bindings {
+        return false;
+    }
+
+    // Check whether any possibly "unknown" patterns are included,
+    // because users might not which values some enum has.
+    // Well-known enums are excepted, as we assume people know them.
+    // We do a deep check, to be able to disallow Err(En::Foo(_))
+    // for usage of the En::Foo variant, as we disallow En::Foo(_),
+    // but we allow Err(_).
```

Comment:
Yes, from a "can this be expressed as a let else" point of view you are right, but there is also the "should this be a let else" question.

I restricted this it as a response to feedback from reviews on my PRs on rustc to adopt `let else`: [first](https://github.com/rust-lang/rust/pull/94146#discussion_r810435895), [second](https://github.com/rust-lang/rust/pull/94146#discussion_r810415177), [third](https://github.com/rust-lang/rust/pull/94146#discussion_r810435895), [fourth](https://github.com/rust-lang/rust/pull/94143#discussion_r810415707), [fifth](https://github.com/rust-lang/rust/pull/94142#discussion_r810490398). See the discussion there, people were concerned that the code would be made less clear if switched to a `match`.

Maybe it might make sense to add a config variable to allow people to tune this. Maybe one should even add an option to make it not lint on `match` at all, as some people prefer `match` over `if let` (due to this, `single_match_else` is a pedantic lint from what I can tell, and not a style lint).

Reply:
Then I think a config option, that can disable this restriction would be the best option. You can, but don't have to, add it in this PR or leave it for a future improvement. If you decide to do the latter, please leave a `FIXME` here, that this should be made configurable. (Maybe include a note about `non_exhaustive`, because we should never lint those)

Reply:
> (Maybe include a note about `non_exhaustive`, because we should never lint those)

Why should matches on `non_exhaustive` enums not be linted?

Reply:
Oh nvm, a `non_exhaustive` enum needs a `_` arm outside of the defining crate, making this lint not apply anyway (if it has more than 1 non-`_` other arm) . I always forget about this.

SO forget what I said about `non_exhaustive`. That doesn't apply to this lint.

Reply:
I've added a config option, to both make the lint fire more often but also to make it fire less often.